### PR TITLE
Fix nil pointer error for Network

### PR DESCRIPTION
**Pull request by rustycl0ck**
**What this PR does / why we need it**: When the network element does not exist (when trying to delete a failed cluster creation), the controller dies with the nil pointer exception. This PR fixes that.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by jichenjc** [comment]
@hidekazuna can you help to provide some comments? Thanks

File: controllers/openstackcluster_controller.go [hunk 1]
```diff
@@ -157,17 +157,19 @@ func (r *OpenStackClusterReconciler) reconcileDelete(ctx context.Context, log lo
 	}
 
 	if openStackCluster.Spec.ManagedAPIServerLoadBalancer {
-		if apiLb := openStackCluster.Status.Network.APIServerLoadBalancer; apiLb != nil {
-			if err = loadBalancerService.DeleteLoadBalancer(apiLb.Name, openStackCluster); err != nil {
-				return reconcile.Result{}, errors.Errorf("failed to delete load balancer: %v", err)
-			}
-			r.Recorder.Eventf(openStackCluster, corev1.EventTypeNormal, "SuccessfulDeleteLoadBalancer", "Deleted load balancer %s with id %s", apiLb.Name, apiLb.ID)
+		if openStackCluster.Status.Network != nil {
```
Comment (hidekazuna) [inline controllers/openstackcluster_controller.go:160]: if openStackCluster.Spec.ManagedAPIServerLoadBalancer && openStackCluster.Status.Network != nil

File: controllers/openstackcluster_controller.go [hunk 2]
```diff
@@ -189,13 +191,15 @@ func (r *OpenStackClusterReconciler) reconcileDelete(ctx context.Context, log lo
 		r.Recorder.Eventf(openStackCluster, corev1.EventTypeNormal, "SuccessfulDeleteSecurityGroup", "Deleted security group %s with id %s", controlPlaneSecGroup.Name, controlPlaneSecGroup.ID)
 	}
 
-	if router := openStackCluster.Status.Network.Router; router != nil {
-		log.Info("Deleting router", "name", router.Name)
-		if err = networkingService.DeleteRouter(openStackCluster.Status.Network); err != nil {
-			return ctrl.Result{}, errors.Errorf("failed to delete router: %v", err)
+	if openStackCluster.Status.Network != nil {
```
Comment (hidekazuna) [inline controllers/openstackcluster_controller.go:194]: Can you merge L194-202 with L206-L213?

File: controllers/openstackcluster_controller.go (hunk 1 above)
Comment (rustycl0ck) [reply controllers/openstackcluster_controller.go]: Fixed!

File: controllers/openstackcluster_controller.go (hunk 2 above)
Comment (rustycl0ck) [reply controllers/openstackcluster_controller.go]: Fixed!

**Comment by rustycl0ck** [comment]
@hidekazuna san, I made the changes you requested. Please review at your convenience.

**Comment by jichenjc** [comment]
recheck

File: controllers/openstackcluster_controller.go [hunk 3]
```diff
@@ -189,17 +189,17 @@ func (r *OpenStackClusterReconciler) reconcileDelete(ctx context.Context, log lo
...
-		if err = networkingService.DeleteRouter(openStackCluster.Status.Network); err != nil {
-			return ctrl.Result{}, errors.Errorf("failed to delete router: %v", err)
-		}
-		r.Recorder.Eventf(openStackCluster, corev1.EventTypeNormal, "SuccessfulDeleteRouter", "Deleted router %s with id %s", router.Name, router.ID)
-		log.Info("OpenStack router deleted successfully")
-	}
-
 	// if NodeCIDR was not set, no network was created.
 	if network := openStackCluster.Status.Network; network != nil && openStackCluster.Spec.NodeCIDR != "" {
+		if router := network.Router; router != nil {
```
Comment (hidekazuna) [inline controllers/openstackcluster_controller.go:194]: Even if openStackCluster.Spec.NodeCIDR == "", router should be deleted;

```go
if network := openStackCluster.Status.Network; network != nil {
  if router := network.Router; router != nil {
    // delete router
  }
  if openStackCluster.Spec.NodeCIDR != "" {
    // delete network
  }
}
```

File: controllers/openstackcluster_controller.go (hunk 1 above)
Comment (hidekazuna) [reply controllers/openstackcluster_controller.go]: Thanks, confirmed.

File: controllers/openstackcluster_controller.go (hunk 3 above)
Comment (rustycl0ck) [reply controllers/openstackcluster_controller.go:194]: From what I understood ([Ref](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/controllers/openstackcluster_controller.go#L402)), the router is created only when `openStackCluster.Spec.NodeCIDR != ""`. Thus we should be deleting the router only in that scenario.
Please correct me if my understanding is wrong.

**Comment by jichenjc** [comment]
recheck

File: controllers/openstackcluster_controller.go (hunk 3 above)
Comment (hidekazuna) [reply controllers/openstackcluster_controller.go:194]: I was wrong, you are right. 
I understand you fix to guard openStackCluster.Status.Network, and openStackCluster.Spec.NodeCIDR != "".

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
we really need talk to open lab about the test env ...

**Comment by jichenjc** [comment]
recheck

**Comment by sbueringer** [comment]
@jichenjc I tried for literally months without any success. But feel free to continue: https://github.com/theopenlab/openlab/issues/141

**Comment by jichenjc** [comment]
@hidekazuna your call :)

**Pull request merged**